src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs

```diff
@@ -687,5 +689,7 @@ public virtual void PrintMessageNormalized(TextWriter writer, string message, bo
             else if (!removeLastNewLine)
                 writer.WriteLine();
         }
+
+        public virtual CompositionContainer GetCompositionContainer() => throw this.ExceptNotSupp("Composition not supported by this environment");
```

Comment:
>public virtual CompositionContainer GetCompositionContainer() => throw this.ExceptNotSupp("Composition not supported by this environment"); [](start = 8, length = 139)

In the event that someone hasn't set up a composition via the local environment, my expectation would have been that a call to this method -- after all part of the interface we claim to be implementing -- would not *fail*, but would instead return some value (an empty container, or even `null`), to indicate that no components had been registered. That we add a method to an interface and then in the base class we throw with this exception is I think a bad sign.